src/table/utils.cc

```diff
@@ -0,0 +1,164 @@
+// -*- lsst-c++ -*-
+
+/*
+ * LSST Data Management System
+ * Copyright 2018 LSST Corporation.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#include <memory>
+#include <vector>
+
+#include "lsst/pex/exceptions.h"
+#include "lsst/afw/table/fwd.h"
+#include "lsst/afw/table/aggregates.h"
+#include "lsst/afw/table/Simple.h"
+#include "lsst/afw/table/Source.h"
+#include "lsst/afw/table/utils.h"
+
+namespace lsst {
+namespace afw {
+namespace table {
+namespace {
+
+/*
+ * Get the schema from a collection of records
+ *
+ * This function and getValues and setValues below appear in two forms:
+ * a catalog of records and a vector of shared pointers to records; the crucial difference
+ * is whether elements are records or shared pointers to records
+ */
+template <typename Record>
+Schema getSchema(SortedCatalogT<Record> const &catalog) {
+    return catalog[0].getSchema();
+}
+
+template <typename Record>
+Schema getSchema(std::vector<std::shared_ptr<Record>> const &recordList) {
+    return recordList[0]->getSchema();
+}
+
+// Get values for a field in a collection of catalog records
+template <typename Value, typename Record, typename Key>
+std::vector<Value> getValues(std::vector<std::shared_ptr<Record>> const &recordList, Key const &key) {
+    std::vector<Value> valueList;
+    valueList.reserve(recordList.size());
+    for (auto const &recordPtr : recordList) {
+        valueList.emplace_back(recordPtr->get(key));
+    }
+    return valueList;
+}
+
+template <typename Value, typename Record, typename Key>
+std::vector<Value> getValues(SortedCatalogT<Record> const &catalog, Key const &key) {
```

Comment:
I think you could expand the types you support a bit by just using `CatalogT<Record>` instead of `SortedCatalogT<Record>` here (and the same for `setValue`, of course), because `SortedCatalogT` inherits from `CatalogT`.

Reply:
Sounds good. When I did this I found that I could no longer rely on the macros to do the right thing so I stopped using the macro. The result is fewer lines of code and that code looks more like the pybind11 file.